File: 2025_Oct/Anica/week2/2025-10-27-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False 

        start, end = 0, len(s1)-1
        count_s1 = [0]*26
        count_s2 = [0]*26 
        
        for i in range(len(s1)):
            count_s1[ord(s1[i])-ord('a')]+=1
            count_s2[ord(s2[i])-ord('a')]+=1

        if count_s1 == count_s2: 
            return True 
            
        end+=1
        while end < len(s2):
            count_s2[ord(s2[end]) - ord('a')]+=1 
            count_s2[ord(s2[start])-ord('a')]-=1
            if count_s1 == count_s2:
                return True 
            start+=1
            end+=1
            

        return False        

File: 2025_Oct/evelyn/week2/2025-10-27-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();

        if (n > m) return false;

        vector<int> count1(26, 0);
        vector<int> count2(26, 0);

        for (int i = 0; i < n; i++) {
            count1[s1[i]-'a']++;
            count2[s2[i]-'a']++;
        }

        for (int i = 0; i < m-n; i++) {
            if (count1==count2) return true;

            count2[s2[i+n]-'a']++;
            count2[s2[i]-'a']--;
        }

        return count1==count2;
    }
};